### ISSUE TITLE
Start Session for an Authenticated User

### ISSUE THREAD

**Issue by JeremyHutchings** [UnderRatedBrilliance/XenforoBridge#8]
Hi,

Any plans to add a facade. I'm trying to figure out how to use this with in Laravel to access and manipulate xF users, though can't find any example or demo code any where.

Thanks
J

**Comment by gabarba** [comment]
I saw your post on XF forums what exactly do you want to manipulate? Creating a Facade should be trivial I can do it when I get a chance either today or tomorrow. But I tend to avoid using facades personally and just inject the XenforoBridge Object directly into the controller or repository that would have a need to access it. I mostly use it as a middleware to start up a xf session and to check if it was authenticated. If not I redirect to the xenforo log in screen. 

The other use case for me is I use the bridge to render my laravel data using the xenforo system.

So my question for you is do you want a facade or do you just want example code for how to use the bridge?

**Comment by JeremyHutchings** [comment]
> I saw your post on XF forums what exactly do you want to manipulate?

We have an existing site built in Laravel, and are thinking about adding XF to it. Given we already have all the user registration already done, I just need a way to manage XF sessions, so I can create one when they log in via the current method. So when navigating to a part of the site that is XF they already have their XF session cookie, and when they log (via Laravel) out we'll kill both sessions. 

With some test / hack code creating a user - and matching on email - was easy enough, just looking at the session handling now.

>  I mostly use it as a middleware to start up a xf session and to check if it was authenticated. If not I redirect to the xenforo log in screen.

Basically what I want do to, though I don't want to redirect, I just want to automatically create the session. Can do the usual if ( ! user exsists ) { create } then login.

**Comment by gabarba** [comment]
This particular bridge is not currently designed set a users session to a particular user_id. I tried to look into how I can modify it to do it quickly myself just now but upon further investigation I would need to put in a bit more thought before I can get you an answer.

 I like the Idea of being able to do that though, because its additional use case could also be used by an admin to login as a particular user. If that was needed by the bridge user. I'll try to spend some time looking into this hopefully this week.

**Comment by JeremyHutchings** [comment]
>  I'll try to spend some time looking into this hopefully this week.

If you could give me any pointers of where to look / start for generating xF sessions I could fork and have a go locally myself :) 

I've started with a stand alone service in Laravel, though would much rather contribute to this project and build out a robust Facade that would allow me to manage the users between both systems.

**Comment by gabarba** [comment]
If you are constraint on time and don't want to wait for a solution from me. You can  take a look at 

https://github.com/VinceG/xenforo-sdk/blob/master/XenForoSDK.php

This is where some of my code is original based from. I see from his code he has a login method that looks like its just a matter of authenticating and changing the session to a particular user model at first glance. 

I would start from line 266 which is the validateLogin() method and work your way down from there. I just didn't have a need for this method so it was never implemented.

**Comment by JeremyHutchings** [comment]
Thanks, will check it out asap.

**Comment by gabarba** [comment]
I have a few other things to take care of today while I'm at work so I wont be able to look at this right this moment. Like I said I'll try to take a look at this  ASAP when I have time.

**Comment by gabarba** [comment]
Sorry I wasn't able to look at this sooner just has been a bit hectic at work as of late. 

I have added a new method loginAsUser() which will allow you to set the session as a particular user. Let me know if this works for you. I have also created a facade class so so you can implement that how ever you would like in your installation.

**Comment by JeremyHutchings** [comment]
Thanks.

Am testing now.

**Comment by JeremyHutchings** [comment]
Just getting back to this now.

> I have added a new method loginAsUser() which will allow you to set the session as a particular user

Where is that and how to call it ? 

>  I have also created a facade class so so you can implement that how ever you would like in your installation.

What are you using for the name space for that ? 

Also the readme.md lists : 

> 'XenforoBridge\XenforoBridgeServiceProvider::class',

It should be 'XenforoBridge\XenforoBridgeServiceProvider', for app/config.php I think

**Comment by JeremyHutchings** [comment]
What alias did you use to test this ? 

```
 'aliases' => [
        'XF'        => 'Urb\XenforoBridge\Facades\XenforoBridge',
    ],
```

**Comment by JeremyHutchings** [comment]
Any chance you can update the main readme - as per https://github.com/UnderRatedBrilliance/XenforoBridge/issues/10 - then I can try and test this as I've not been able to get anything to work so far.

**Comment by gabarba** [comment]
Sorry I wont be able to update documentation till after the holidays.

You can use what ever alias you would like, I believe I also used XF in testing it. Also the method loginAsUser()  is located in src/XenforoBridge/XenforoBrige.php on line 421.

**Comment by JeremyHutchings** [comment]
Did you use an alias to test this ? 

I can't figure out what path to the namespace it should be as it looks like it has conflicting names at the moment.

**Comment by JeremyHutchings** [comment]
I manage to get some of it started with : 

``` php
'providers' => [
    'XenforoBridge\XenforoBridgeServiceProvider',
],
```

and

``` php
'aliases' => [
        'XF' => 'XenforoBridge\Facades\XenforoBridge',
    ],
```

Though this : 

``` php
\XF::loginAsUser(1);
```

Throws an error : 
`Cannot redeclare class XenForo_Autoloader`

As it's reloading it, so I don't think this code has been tested with a Facade. 

What code did you use to test or get a login from Laravel working? Did you use a require() some where in a controller and instantiate the class directly?

**Comment by JeremyHutchings** [comment]
If you could explain any way that loginAsUser() can be used or accessed it might help me get going on this, because at the moment I cant see any way to get at it without the service loading twice.

**Comment by gabarba** [comment]
There is something wrong apparently. It is suppose to only load once, by
design it uses a shared object or singleton.

**Comment by JeremyHutchings** [comment]
How did you access the singleton when testing?

**Comment by gabarba** [comment]
I normally inject the class I dont use the facades in most of my applications particular this one that it was originally developed. I'm pretty busy tonight a few things due on the 10th and another feature to add to our site before I leave on the 15th for vacation and wont be back till the 28th. Is there any way you can send me a sample of your code or access to view a repo? It seems like you are initiating the Xenforo_Autoloader at some other point other than with my service provider. Are you using the middleware as well? If I have time later tonight I'll try to look at it after I am done working my other project. I wish you had messaged me sooner not exactly the best timing but I'll try to help you best I can.

**Comment by JeremyHutchings** [comment]
What might be best to ignore what I've done for now, and you share with me how you'd use the code to login a user ? And create one (if it dose that atm?)

**Comment by JeremyHutchings** [comment]
> I normally inject the class I dont use the facades in most of my applications particular this one that it was originally developed. 

Into the class itself using IoC ?

> Is there any way you can send me a sample of your code or access to view a repo? 

There isn't any code at the moment as I'm just trying to get access to the bridge for the loginAsUser() to be callable. 

> It seems like you are initiating the Xenforo_Autoloader at some other point other than with my service provider. 

Yes, I was using $xF = new class to try and get at the function, though it's already auto loaded, though I couldn't access "xenforobrige" in the global namespace, or from the $app.

> Are you using the middleware as well? 

No,

> I wish you had messaged me sooner not exactly the best timing but I'll try to help you best I can.

I know me to, though projects and time line ! :)

**Comment by JeremyHutchings** [comment]
> and wont be back till the 28th

Ping :)

**Comment by JeremyHutchings** [comment]
Hi,

Just wondering if you are back and and can share how you use this code in a Laravel controller to create a session (and maybe a user if they don't exist) ?

Cheers

**Comment by JeremyHutchings** [comment]
Hello. Do you have any time at the moment to post an example of how you're using the code ? Thanks.

**Comment by gabarba** [comment]
Sorry man, I've been really swamped with work since coming back. Making up for 3 week being gone and need to catch up on Bills. I would look over the execution and see at what points you are calling the Xenforo Autoloader. 

I just pushed a commit to include the XenForo_Autoloader once. This hopefully should resolve the 

`Cannot redeclare class XenForo_Autoloader` 

Error you are receiving

**Comment by JeremyHutchings** [comment]
Just working though :
```
No entry is registered for key 'session'
```
At the moment, looks like Zend config.

**Comment by gabarba** [comment]
So did this resolve your error? Can I close this issue?

**Comment by JeremyHutchings** [comment]
> So did this resolve your error? Can I close this issue?

Any chance you can just show me the controller code you are using to test the lib ?

I'm just trying to check if a user exsists in xF dB, if they don't create them them log them in.

**Comment by JeremyHutchings** [comment]
Any chance you can just show me the controller code you are using to test the lib ?

**Comment by gabarba** [comment]
Here is a stripped dowm example of one my controllers using XenforoBridge for Session Authentication. 

``` php
<?php

namespace App\ItemHub\Item\Controllers;

use App\Http\Controllers\Controller;
use App\ItemHub\Item\Contracts\ItemRepositoryInterface;
use App\ItemHub\Review\Contracts\ReviewRepositoryInterface;
use XenforoBridge\XenforoBridge;
use App\ItemHub\Item\Requests\FrontendStoreReview;

class ItemReviewFrontendController extends Controller{

    protected $xenforo;

    protected $item;

    public function __construct(ItemRepositoryInterface $item,XenforoBridge $xenforo)
    {
        /**
         * Ensure that user is logged in to create or store reviews
         */
        $this->middleware('xen.auth',['only' =>['create','store']]);
        $this->item   = $item;
        $this->xenforo = $xenforo;
    }

    /**
     * Create form to submit review. Will not display page if does not pass validation
     * from FrontendStoreReview
     *
     * @param $slug
     * @param FrontendStoreReview $request
     * @return mixed
     */
    public function create($slug, FrontendStoreReview $request)
    {
        $item = $this->item->findBySlug($slug);


        $view = view('Item.Views.Reviews.Form',['items' => $item]);

        return $this->xenforo->renderTemplate('urb_itemhub_review_form', '',$this->transformCreate($item->toArray(),(string) $view));
    }

    /**
     *
     * Store Review after it passes route validation from FrontendStoreReview
     * Inject Current User_id and ItemId
     *
     * @todo need to trigger event beforeReviewCreation and afterReviewCreation
     * @todo need to create command that triggers recalculating reviewCode Stats and reindex ItemFlat
     *
     * @param $slug
     * @param FrontendStoreReview $request
     * @param ReviewRepositoryInterface $review
     * @return \Illuminate\Http\RedirectResponse
     */
    public function store($slug, FrontendStoreReview $request, ReviewRepositoryInterface $review)
    {


    }

}

```

Basically the part that does all the authentication is the middleware before it even hits a controller method. You can see I have set the middleware on only 2 methods in the class. I also inject the XenforoBridge because I use it specifically in the create method where I send a view to be injected in an existing xenforo template and rendered via the Xenforo templating engine.  There is quite a bit more to this but controller but that is the gist of it in use.

**Comment by gabarba** [comment]
The other exception you are experiencing sounds like a missing configuration or improper installation.

**Comment by JeremyHutchings** [comment]
In my app, the Laravel has to control the xF user so it's more a case of : 

```
$laravelUser = \Auth::user();
$xenforoUser = XF::getUserByEmail($laravelUser->email);

if ($xenforoUser)
{
    $test = XF::loginAsUser($xFuser);
    dd($test);
}
else
{
    // Create the xF user from the Laravel one, and log them in
}
```

I get an array for $xenforoUser of the expeted user, though dies with no response trying to log them in at the moment.

**Comment by JeremyHutchings** [comment]
Using XF::loginAsUser($user, true, false) to get this line to be called : 

```
$this->getXenforoModel('XenForo_Model_User')->setUserRememberCookie($user);
```

gives a param to bind error

```
mysqli_stmt::bind_param(): Number of variables doesn't match number of parameters in prepared statement
```

Looks like the User Model in XF can't get the userId from what is passed in.

Is XF::getUserByEmail supposed to return an array or an object ?

**Comment by JeremyHutchings** [comment]
Looks like it's  this that confused me : 

```
    public function loginAsUser($user, $remember = false, $log = true)
```

$user opposed to $userId.

Once it's the ID of the xF user passed in, so 

```
$test = XF::loginAsUser($xFuser['user_id']);
dd($test);
```

Opposed to : 

```
$test = XF::loginAsUser($xFuser);
dd($test);
```

The code runs and I can see a cookie `xf_user=`  being set .... though not logged in yet.

**Comment by gabarba** [comment]
Yes it suppose to return an array if no user is returned it returns an empty array. Also what php version are you working on?

**Comment by JeremyHutchings** [comment]
5.6.11 at the moment.

**Comment by gabarba** [comment]
So have you tried to log in using an existing user_id ?

**Comment by gabarba** [comment]
you may have to do this if I remember correctly It didn't hold the authentication unless you set remember to true like so, 

``` php

XF::loginAsUser($xFuser,true);
```

Not sure why it is but I remember it only authenticated for that request and was not persistent unless I set that as true.

**Comment by JeremyHutchings** [comment]
Yes, I've created a user in xF with the same email as my auth'ed user in Laravel, so that the look up finds it (will deal with creating the users later). so when I look up with my email, I get user_id 3, from xF.

Then trying to log that user in to xF at the moment : 

```
            $user   = \Auth::user();
            $xFuser = XF::getUserByEmail($user->email);

            XF::loginAsUser($xFuser['user_id'], true, false);
```

$xFuser['user_id'] is 3 there.

**Comment by gabarba** [comment]
when you goto the forums it does not have you logged in?

**Comment by JeremyHutchings** [comment]
I can see a `xf_user` user cookie coming back from the Laravel page where the code is running.

Though I don't see it being passed when I go to the forums, though I see a `xf_session` cookie.

> when you goto the forums it does not have you logged in?

Nope, that is what I'm trying to diagnose atm.

**Comment by gabarba** [comment]
What domain is your forums configured to? does it share the same domain or are you using a sub domain for the forums?

**Comment by JeremyHutchings** [comment]
https://local.mytest.com and http://local.forum.mytest.com all local.

I might spend some time setting up the local dev forum under httpS to see if that is it. But I don't think so.

**Comment by gabarba** [comment]
I would check your cookies between the 2 sites. I suspect because they are being set to 2 different domains that is why Xenforo would not be able to read it.

**Comment by JeremyHutchings** [comment]
The cookie domain is set by the xF config, not the Laravel, so it would have to be correct. 

I can see the xf_user cookie being created by Laravel and passed over now, though I don't see it creating a xf_session one, and the cookie domains are correct and both sites on ssl now has the cookie is http secure.

**Comment by gabarba** [comment]
thats weird because the xf_session is started within the constructor of the XenforoBridge class. Thats what the bootstrapXenforo method does.

``` php
  /**
     * Bootstrap Xenforo Application and Start a Public Session
     *
     * @param string $directoryPath
     */
    protected function bootstrapXenforo($directoryPath)
    {
        XenForo_Autoloader::getInstance()->setupAutoloader($directoryPath .'/library');
        XenForo_Application::initialize($directoryPath . '/library', $directoryPath);
        XenForo_Session::startPublicSession();
    }
```

**Comment by JeremyHutchings** [comment]
I might be getting some where here ! :)

1) 
What would be the easy way to log a user out, i.e the opposite of loginAsUser() 

As when someone logs out of Laravel, we'll have to control the xF session from there and kill it.

2)
Best way to create a user in xF from the Laravel side ?

**Comment by gabarba** [comment]
To be honest didn't even look into that yet but just checked it out you can do it by the following code 

``` php

XF::getSession()->delete()

```

This is according to the XenForo_Session class

**Comment by JeremyHutchings** [comment]
That's returning a null for me at the moment.

**Comment by gabarba** [comment]
Yea looking at the method in the XenForo_Session i has no return value. But looks like it will delete the cookie  and remove the session.

**Comment by JeremyHutchings** [comment]
> XF::getSession()->delete()

That isn't doing it yet. Still investigating.

**Comment by JeremyHutchings** [comment]
Had to manually clear the cookies and call logout in Laravel.

```
$this->auth->logout();

XF::getSession()->delete();
$cookie = \Cookie::forget('xf_user', 'xf_session');

return \Redirect::to('/')->withCookie($cookie);
```

**Comment by JeremyHutchings** [comment]
OK .... I _think_ we can close this one now. I'm going to carry on testing.
